### DISCUSSION TITLE
Create a new theme for UData

### DISCUSSION THREAD

**Discussion by sylvainbeo** [opendatateam/udata#2964] [Q&A] [answered]
Hi there,

I'm Sylvain, working at Oslandia.

I'm currently trying to create / customise a theme for UData. I've been following the [documentation](https://udata.readthedocs.io/en/dev/creating-theme/). Thanks for that. 

I'm in the case where i want to just customize some style and colors  (mark as very easy). but i'm not really sure what to do :

1. I tried what it is recommended :

```
pip install cookiecutter
cookiecutter gh:opendatateam/cookiecutter-udata-theme
```
But it gives me a very basic theme, which is not really working (missing templates etc...). 

2. I tried to copy the gouvfr theme, but it's a huge work to check every files which contains references to gouvfr.

What is the good way to go ? 

Thank you

Sylvain

**Comment by nicolaskempf57** [comment]
Hi Sylvain,

It seems that you are following a previous version of the documentation and not the [current one](https://udata.readthedocs.io/en/stable/creating-theme/).

Due to changes in the front-end architecture and the use of the French government's design system, we don't provide an easy way to only customize style and colors anymore so it will require some more work on your side.

Let me know if you have more questions after going through the documentation.

By the way, where did you find the link to the version of the documentation you are currently following ?

Best regards,
Nicolas

**Reply by sylvainbeo** [reply]
Hi Nicolas,

> It seems that you are following a previous version of the documentation and not the current one.

All right. I followed the current one, but i was hopping the one i used was a newest (in my mind dev is always newer than stable). I don"t remember where i found it. Anyway, on the documentation you pointed me, i guess i have to copy the theme gouvfr, et make the changes i want by just editing the CSS/HTML/JS files ?

> we don't provide an easy way to only customize style and colors anymore

The documentation says : "start from gouvfr and modify some style and colors ⇨ easy". But i guess the documentation is out to date so.

Thank you the informations.

Sylvain.

**Reply by nicolaskempf57** [reply, marked as answer]
I'll have a look at readthedoc to see if we can remove this `dev` version because it's not linked to any commit anymore and documents a 2017 version of the project.

> Anyway, on the documentation you pointed me, i guess i have to copy the theme gouvfr, et make the changes i want by just editing the CSS/HTML/JS files ?

Sure, as stated in the documentation, if you don't want to make your own independant theme, you can fork udata-front and use the `gouvfr` theme as your starting point. Then, you can edit the LESS/Jinja/TS files to suit your needs.

You are free to modify your fork of the `gouvfr` theme as long as you follow the licenses and it's quite easy for a developer but we don't provide an easy way to customize it.

**Reply by sylvainbeo** [reply]
@nicolaskempf57 Sorry for the delay. Thank you for your last answer.

**Reply by sylvainbeo** [reply]
@nicolaskempf57 A few more question if you don't mind.

- When using the docker compose installation, how do you use the udata admin commands (https://udata.readthedocs.io/en/stable/administrative-tasks/).  The docker compose is working well, but how can i create a new user to log in ?

- Back to my theme customisation, again if  use the docker compose installation, may i configure the udata-front somewhere so it takes my fork instead of the official one ?


Thank you

**Reply by sylvainbeo** [reply]
Ok i've got the answer to my first question.

I did : `docker exec -it docker-udata_udata_1 udata user create`

But the second one remains :)

**Reply by nicolaskempf57** [reply]
Are you using this [docker-compose](https://github.com/opendatateam/docker-udata/blob/master/docker-compose.yml) ?

I'll look into it to update the documentation about theme in docker-udata.

**Reply by sylvainbeo** [reply]
Yes i'm using this one.

**Reply by sylvainbeo** [reply]
To be more precise, a copy of that one (in my fork)

**Reply by nicolaskempf57** [reply]
I did the exact same setup to better answer your question.
I made forks of both docker-udata and udata-front.

I made 3 changes in docker-udata :
1. update the requirements
2. add the theme volume
3. fix the entrypoint and docker command

## Update the requirements

`udata-front` can be removed from the file because the local one will be used.

The `requirements.pip` file is outdated and new versions are available:

```
udata==7.0.4
udata-ckan==3.0.2
udata-geoplatform==2.0.0
udata-ods==4.0.0
udata-piwik==4.0.0
udata-recommendations==3.1.5
```

### Add the theme volume

A new volume is necessary for the udata service in `docker-compose.yml` and should point to the local folder of your udata-front fork:

```
- ../udata-front:/src/udata-front
```

### Fix the entrypoint and docker command

To have a working environment, I had to add a command to the udata service in docker-compose :

```
command: serve --host 0.0.0.0 --debugger --reload
```

Also, I had to remove this part from the `entrypoint.sh` file:

```
for r in /src/*.pip ; do
    echo "Installing dependencies from $r"
    pip install -r "$d"
done
```

### Development

You can follow the udata-front JavaScript documentation to install dependencies and run npm scripts.

**Reply by sylvainbeo** [reply]
@nicolaskempf57 Thank you for taking time for me.

I did the changes.

When i launch my **docker compose up**, all is working. 

But if i change something in the udata-front, nothing happen in my browser (i tried to change a translation, and to remove the logo for example). 

I tried with the **npm start** launched in the udata-front directory, but also without.

Is there anything to change also in the udata.cfg file ?  I tried to comment  **PLUGINS = ['front']** and **THEME = 'gouvfr'**. But it doesnt seem to change anything. the interface works well with the gouvfr theme.


Thank you

Sylvain

**Reply by sylvainbeo** [reply]
-- ERRATUM --

It's working. I had a mistake in my docker file.

Thank you.

I'll tell you if i manage to do what i want..

**Reply by Guts** [reply]
Thanks you, it's really helpful :pray:!

**Reply by sylvainbeo** [reply]
@nicolaskempf57 One more question for tonight, with this method, the translation does not seem to work anymore, did you notice that ?

**Reply by nicolaskempf57** [reply]
Translations are pre-compiled in udata-front on pypi. With your own fork, this is not the case.

To compile translations, you must install develop requirements :
- use docker to get a bash inside your container, `docker exec -it docker-udata_udata_1 bash`
- move to your plugin folder
- install develop requirements, `pip install -r requirements/develop.pip `

Then, you can compile your translations with `inv i18nc`.

**Reply by nicolaskempf57** [reply]
Also, as you're using your image as a development environment, you should probably disable cache by using `CACHE_TYPE = 'null'` in `udata.cfg`.

**Reply by sylvainbeo** [reply]
@nicolaskempf57 Hi Nicolas.

I was working on my UData today and i found out that the admin page does not work anymore.

In the console, i've got :404: not found on :

http://localhost:7000/static/admin.css?_=7.0.4
http://localhost:7000/static/admin.js?_=7.0.4
http://localhost:7000/static/common.js?_=7.0.4 

I tried to relbuild the Theme (npm run build --prod), but when i check in the static repository, i found for example  admin.3.5.4.dev0.js

Any clue ? 

Thank you

**Reply by nicolaskempf57** [reply]
The admin at `http://localhost:7000/en/admin/` is coming from udata and is not related to udata-front or files from it. I'm reproducing this too so I'll make a ticket here.

**Reply by sylvainbeo** [reply]
@nicolaskempf57 OK.

**Reply by nicolaskempf57** [reply]
@sylvainbeo you can add the [missing environment variable](https://github.com/opendatateam/docker-udata/pull/347/files#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3) in the docker-compose file :

```
    environment:
      - FLASK_DEBUG=true
```

**Reply by sylvainbeo** [reply]
It's working. Thank you

**Reply by sylvainbeo** [reply]
Hey @nicolaskempf57 . Is there a simple way to change the default port 7000 to another for the front ?

**Reply by nicolaskempf57** [reply]
You can add a `--port` to the udata service `command` in docker compose and change the docker compose `port`  too

**Reply by sylvainbeo** [reply]
Thanks a lot
